[spark] Replace create existing tag semantic with replace_tag

askwang opened the pull request:
<!-- Please specify the module before the PR name: [core] ... or [flink] ... -->

### Purpose
In default way, the create_tag procedure for an existing tag means to update it or fail,  it's semantic is not very accurate, replace it with repalce_tag procedure.

before semantic
```sql
CALL paimon.sys.create_tag(table => 'test.T', tag => 'test_tag', snapshot => 1)")

-- success. update it
CALL paimon.sys.create_tag(table => 'test.T', tag => 'test_tag', snapshot => 1, time_retained => '1 d')")

-- failed
CALL paimon.sys.create_tag(table => 'test.T', tag => 'test_tag', snapshot => 2)")
```

after semantic
```sql
CALL paimon.sys.create_tag(table => 'test.T', tag => 'test_tag', snapshot => 1)")

-- failed
CALL paimon.sys.create_tag(table => 'test.T', tag => 'test_tag', snapshot => 1, time_retained => '1 d')")

-- success. use replace to update
CALL paimon.sys.replace_tag(table => 'test.T', tag => 'test_tag', snapshot => 2, time_retained => '1 d')")
```

<!-- What is the purpose of the change -->

### Tests

<!-- List UT and IT cases to verify this change -->

### API and Format

<!-- Does this change affect API or storage format -->

### Documentation

<!-- Does this change introduce a new feature -->

wwj6591812 commented on docs/content/spark/procedures.md:
```diff
@@ -121,6 +121,19 @@ This section introduce all available spark procedures about paimon.
       </td>
     </tr>
+    <tr>
+      <td>replace_tag</td>
```
I think update_tag may be better?

wwj6591812 commented on docs/content/spark/procedures.md:
```diff
@@ -121,6 +121,19 @@ This section introduce all available spark procedures about paimon.
+         Replace an existing tag with new tag info. Arguments:
+            <li>table: the target table identifier. Cannot be empty.</li>
+            <li>tag: name of the existed tag. Cannot be empty.</li>
+            <li>snapshot(Long):  id of the snapshot which the tag is based on.</li>
```
1、snapshotId
2、It is optional.

wwj6591812 commented on docs/content/spark/procedures.md:
```diff
@@ -121,6 +121,19 @@ This section introduce all available spark procedures about paimon.
+            <li>table: the target table identifier. Cannot be empty.</li>
+            <li>tag: name of the existed tag. Cannot be empty.</li>
+            <li>snapshot(Long):  id of the snapshot which the tag is based on.</li>
+            <li>time_retained: The maximum time retained for the existing tag.</li>
```
2、It is optional.

wwj6591812 commented on paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:
```diff
@@ -576,6 +576,25 @@ public void renameTag(String tagName, String targetTagName) {
+    public void replaceTag(String tagName, @Nullable Duration timeRetained) {
+        Snapshot latestSnapshot = snapshotManager().latestSnapshot();
+        SnapshotNotExistException.checkNotNull(
+                latestSnapshot, "Cannot create tag because latest snapshot doesn't exist.");
```
Can not update tag because latest snapshot doesn't exist.

wwj6591812 commented on paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:
```diff
@@ -576,6 +576,25 @@ public void renameTag(String tagName, String targetTagName) {
+                        findSnapshot(fromSnapshotId),
+                        tagName,
+                        timeRetained,
+                        store().createTagCallbacks());
```
Here you should think this case:
If user update a tag, but he don't want call createTagCallbacks.

askwang replied:
good, replace the existing tag is no need to call callback again.

wwj6591812 commented on paimon-core/src/main/java/org/apache/paimon/table/Table.java:
```diff
@@ -120,6 +120,12 @@ default String fullName() {
     void renameTag(String tagName, String targetTagName);
 
+    @Experimental
+    void replaceTag(String tagName, Duration timeRetained);
```
接口方法【replaceTag】必须使用javadoc注释

wwj6591812 commented on paimon-core/src/main/java/org/apache/paimon/utils/TagManager.java:
```diff
@@ -103,7 +103,27 @@ public void createTag(
+        createOrReplaceTag(snapshot, tagName, timeRetained, callbacks);
+    }
 
+    public void createOrReplaceTag(
```
createOrUpdateTag

wwj6591812 commented on paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/CreateTagProcedure.java:
```diff
@@ -18,71 +18,26 @@
-                    });
+public class CreateTagProcedure extends CreateOrReplaceTagBaseProcedure {
 
     protected CreateTagProcedure(TableCatalog tableCatalog) {
```
why not private?

wwj6591812 commented on paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/ReplaceTagTagProcedure.java:
```diff
@@ -0,0 +1,56 @@
+import java.time.Duration;
+
+/** A procedure to replace a tag. */
+public class ReplaceTagTagProcedure extends CreateOrReplaceTagBaseProcedure {
```
UpdateTagProcedure

wwj6591812 commented on paimon-spark/paimon-spark-common/src/main/java/org/apache/paimon/spark/procedure/ReplaceTagTagProcedure.java:
```diff
@@ -0,0 +1,56 @@
+/** A procedure to replace a tag. */
+public class ReplaceTagTagProcedure extends CreateOrReplaceTagBaseProcedure {
+
+    protected ReplaceTagTagProcedure(TableCatalog tableCatalog) {
```
why not private?

wwj6591812 reviewed (commented):
1、I think update_tag may be better?
2、You alse need modify flink action&& procedure and doc.

askwang commented:
addressed comments, thanks for review @wwj6591812 
A little difference, keep the name 'replace_tag' not change. I think it's name is ok, and iceberg also uses 'replace tag' to update tag. what do you think?

wwj6591812 commented:
> addressed comments, thanks for review @wwj6591812 A little difference, keep the name 'replace_tag' not change. I think it's name is ok, and iceberg also uses 'replace tag' to update tag. what do you think?

ok

JingsongLi commented on paimon-core/src/main/java/org/apache/paimon/table/AbstractFileStoreTable.java:
```diff
@@ -576,6 +576,19 @@ public void renameTag(String tagName, String targetTagName) {
+    }
+
+    @Override
+    public void replaceTag(String tagName, long fromSnapshotId, @Nullable Duration timeRetained) {
```
Just create a `replaceTag(String tagName, @Nullable Long fromSnapshotId, @Nullable Duration timeRetained)`?

JingsongLi reviewed (approved):
+1
